src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    /// <summary>
+    /// A singleton that tracks the open IVsWindowFrames and can report which documents are visible or active in a given <see cref="Workspace"/>.
+    /// Can be accessed via the <see cref="IDocumentTrackingService"/> as a workspace service.
+    /// </summary>
+    [Export]
+    internal class VisualStudioActiveDocumentTracker : ForegroundThreadAffinitizedObject, IVsSelectionEvents, IDisposable
+    {
+        private readonly IThreadingContext _threadingContext;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+
+        /// <summary>
+        /// Collection of all asynchronous tasks that are started by this service. This should only be tasks that are implicitly
+        /// async as we fetch services from the <see cref="IAsyncServiceProvider"/>, and are used to ensure we wait for them during shutdown.
+        /// These should not be waited for in calls to <see cref="GetActiveDocument(Workspace)"/> or <see cref="GetVisibleDocuments(Workspace)"/>
+        /// because those are expected to not be jumping to the UI thread per traditional Roslyn semantics and may deadlock.
+        /// </summary>
+        private readonly JoinableTaskCollection _asyncTasks;
+
+        private ImmutableList<FrameListener> _visibleFrames = ImmutableList<FrameListener>.Empty;
+        private IVsWindowFrame _activeFrame;
+
+        [ImportingConstructor]
+        public VisualStudioActiveDocumentTracker(IThreadingContext threadingContext, [Import(typeof(SVsServiceProvider))] IAsyncServiceProvider asyncServiceProvider, IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+            : base(threadingContext, assertIsForeground: false)
+        {
+            _threadingContext = threadingContext;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _asyncTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
+            _asyncTasks.Add(threadingContext.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var monitorSelectionService = (IVsMonitorSelection)await asyncServiceProvider.GetServiceAsync(typeof(SVsShellMonitorSelection));
+
+                if (ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_DocumentFrame, out var value)))
+                {
+                    if (value is IVsWindowFrame windowFrame)
+                    {
+                        TrackWindowFrame(windowFrame);
+                        _activeFrame = windowFrame;
+                    }
+                }
+
+                monitorSelectionService.AdviseSelectionEvents(this, out var _);
```

Comment:
i think it would be nice to document this whole lambda :)

Reply:
What part seemed off? It's fairly standard code.

Reply:
Strange.  This no longer links to the original code...  Give me a bit.

Reply:
For example, why we do the `if (value is IVsWindowFrame)` + tracking here.  It's unclear why we hav ehtat code here, and also in OnElementValueChanged.  it seems like we should share this.  Then this is only dealing with the case where we need to get the initial data, and all future changes are done in the callback.  But both call into the same code to proces sthings.

Reply:
Ah, yeah, makes sense. Fixing.

Reply:
Fixed